src/alpaka/AlpakaCore/alpakaDevices.h

```diff
@@ -11,29 +11,17 @@
 
 namespace cms::alpakatools {
 
-  // alpaka host device
-  inline const alpaka_common::DevHost host = alpaka::getDevByIdx<alpaka_common::PltfHost>(0u);
+  // alpaka host platform and device
+  inline const alpaka_common::PlatformHost platformHost{};
+  inline const alpaka_common::DevHost host = alpaka::getDevByIdx(platformHost, 0u);
 
-  // alpaka accelerator devices
+  // alpaka accelerator platform and devices
+  // these objects are filled by a call to cms::alpakatools::initialise<TPlatform>()
   template <typename TPlatform>
-  inline std::vector<alpaka::Dev<TPlatform>> devices;
+  inline std::optional<TPlatform> platform;
 
   template <typename TPlatform>
-  std::vector<alpaka::Dev<TPlatform>> enumerate() {
-    assert(getDeviceIndex(host) == 0u);
-
-    using Device = alpaka::Dev<TPlatform>;
-    using Platform = TPlatform;
-
-    std::vector<Device> devices;
-    uint32_t n = alpaka::getDevCount<Platform>();
-    devices.reserve(n);
-    for (uint32_t i = 0; i < n; ++i) {
-      devices.push_back(alpaka::getDevByIdx<Platform>(i));
-      assert(getDeviceIndex(devices.back()) == static_cast<int>(i));
-    }
-    return devices;
-  }
+  inline std::vector<alpaka::Dev<TPlatform>> devices;
```

Comment:
Would it make sense (now) to move to the same approach as in CMSSW, i.e. have a function that stores the global object in a function-scope `static` variable and returns a (const?) reference to the object?

Reply:
I think so, but I would need to try and see if anything breaks (and then fix it).

Reply:
Something like
```c++
namespace cms::alpakatools {
      
  // return the alpaka accelerator platform for the given platform
  template <typename TPlatform, typename = std::enable_if_t<alpaka::isPlatform<TPlatform>>>
  inline TPlatform const& platform() {
    static const auto platform = TPlatform{};
    return platform;
  } 
  
  // return the alpaka accelerator devices for the given platform
  template <typename TPlatform, typename = std::enable_if_t<alpaka::isPlatform<TPlatform>>>
  inline std::vector<alpaka::Dev<TPlatform>> const& devices() {
    static const auto devices = alpaka::getDevs(platform());
    return devices;
  }

}  // namespace cms::alpakatools
```
?

I would also take the opportunity to try and make then not `inline`, to avoid problems with linking.

Reply:
An attempt is in https://github.com/cms-patatrack/pixeltrack-standalone/pull/404 .